File: DogeChat/Video/FriendVideoView.m
//
//  FriendVideoView.m
//  DogeChat
//
//  Created by 赵锡光 on 2021/6/17.
//  Copyright © 2021 Luke Parham. All rights reserved.
//

#import "FriendVideoView.h"
#import "DogeChat-Swift.h"
#import <Foundation/Foundation.h>

@implementation FriendVideoView {
    Renderer *_renderer;
}

- (instancetype)initWithFrame:(CGRect)frame {
    self = [super initWithFrame:frame];
    if (self) {
        [self prepare];
    }
    return self;
}

- (void)prepare {
    self.device = MTLCreateSystemDefaultDevice();
    self.backgroundColor = [UIColor clearColor];
    self.colorPixelFormat = MTLPixelFormatBGRA8Unorm;
    self.depthStencilPixelFormat = MTLPixelFormatDepth32Float_Stencil8;
    _renderer = [[Renderer alloc] initWithDevice:self.device renderDestination:self];
    [_renderer mtkView:self drawableSizeWillChange:self.bounds.size];
    self.delegate = _renderer;
}

- (void)didReceiveData:(NSData *)data {
    NSRange range = NSMakeRange(0, 4);
    NSData *headData = [data subdataWithRange:range];
    int head= [self intWithData:headData];
    
    NSRange countRange = NSMakeRange(4,4);
    NSData *countData = [data subdataWithRange:countRange];
    int count = [self intWithData:countData];
    
    NSRange lengthRange = NSMakeRange(8,4);
    NSData *lengthData = [data subdataWithRange:lengthRange];
    int length = [self intWithData:lengthData];
    
    
    NSRange typeRange = NSMakeRange(12,4);
    NSData *typeData = [data subdataWithRange:typeRange];
    int type = [self intWithData:typeData];
    int completeDataLength = length+16;
    //打印字节头  序列号 消息内容长度  这个包总长
    NSLog(@"head=%d count=%d bodylength=%d completeDataLength=%d completeData.length=%lu",head,count,length,completeDataLength,(unsigned long)data.length);
    
    if (data.length>=completeDataLength) {
        //取出body数据
        NSRange dataRange = NSMakeRange(16, length);
        NSData *mData = [data subdataWithRange:dataRange];
        if (type == 0) {
        }else if (type == 1) {
            // 视频
            VideoFrameData *frameData = [NSJSONSerialization object ]
            __weak MainVideoView *wself = self;
        }
        
    }
}

- (int)intWithData:(NSData *)data {
    Byte *byteK = (Byte *)data.bytes;
    int valueK;
    valueK = (int) (((byteK[0] & 0xFF)<<24)
                    |((byteK[1] & 0xFF)<<16)
                    |((byteK[2] & 0xFF)<<8)
                    |(byteK[3] & 0xFF));
    return valueK;
}


- (int)intWithDataBytes:(char *)byteK {
    // Byte *byteK = (Byte *)data.bytes;
    int valueK;
    valueK = (int) (((byteK[0] & 0xFF)<<24)
                    |((byteK[1] & 0xFF)<<16)
                    |((byteK[2] & 0xFF)<<8)
                    |(byteK[3] & 0xFF));
    return valueK;
}


@end

File: DogeChat/Video/FriendVideoView.swift
//
//  FriendVideoView.swift
//  DogeChat
//
//  Created by 赵锡光 on 2021/6/17.
//  Copyright © 2021 Luke Parham. All rights reserved.
//

import UIKit
import MetalKit

@available(iOS 13.0, *)
@objc class FriendVideoView: MTKView {
    
    var renderer: Renderer!
    @objc var processor: VideoProcessor!
    var lastWidthAndHeight: (width: CGFloat, height: CGFloat) = (0, 0)
    
    override init(frame: CGRect, device: MTLDevice?) {
        let width = UIScreen.main.bounds.width
        let center = UIScreen.main.bounds.center
        let height = 640 * width / 360
        super.init(frame: CGRect(center: center, size: CGSize(width: width, height: height)), device: nil)
    }
    
    required init(coder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
    @objc func prepare() {
        guard let device = MTLCreateSystemDefaultDevice() else {
            fatalError("Unable to get system default device!")
        }
        self.device = device
        self.backgroundColor = .clear
        self.colorPixelFormat = .bgra8Unorm
        self.depthStencilPixelFormat = .depth32Float_stencil8
        
        renderer = Renderer(device: device, renderDestination: self)
        renderer.mtkView(self, drawableSizeWillChange: self.bounds.size)
        self.delegate = renderer
        processor = VideoProcessor()
    }
    
    @objc func didReceiveData(_ data: Data) {
        let nsdata = data as NSData
        _ = Recorder.int(with: nsdata.subdata(with: NSRange(location: 0, length: 4)))
        _ = Recorder.int(with: nsdata.subdata(with: NSRange(location: 4, length: 4)))
        let length = Recorder.int(with: nsdata.subdata(with: NSRange(location: 8, length: 4)))
        let type = Recorder.int(with: nsdata.subdata(with: NSRange(location: 12, length: 4)))
        let videoData = nsdata.subdata(with: NSRange(location: 16, length: Int(length)))
        if type == 1 {
            guard let frameData = try? JSONDecoder().decode(VideoFrameData.self, from: videoData) else { return }
            let sampleBuffer = frameData.makeSampleBuffer()
            self.processor.decompress(sampleBuffer) { [self] imageBuffer, presentationTimeStamp in
                let width = CGFloat(CVPixelBufferGetWidth(imageBuffer))
                let height = CGFloat(CVPixelBufferGetHeight(imageBuffer))
                DispatchQueue.main.async {
                    self.updateLayout(width: width, height: height)
                }
                self.renderer.enqueueFrame(pixelBuffer: imageBuffer, presentationTimeStamp: presentationTimeStamp, inverseProjectionMatrix: nil, inverseViewMatrix: nil)
            }
        }
    }
    
    func updateLayout(width: CGFloat, height: CGFloat) {
        guard let view = self.superview else { return }
        if lastWidthAndHeight.width == width && lastWidthAndHeight.height == height {
            return
        }
        lastWidthAndHeight = (width, height)
        let screenWidth = UIScreen.main.bounds.width
        self.frame = CGRect(x: 0, y: 0, width: screenWidth, height: screenWidth * height / width)
        self.center = view.center
    }

}
